evm/src/trace/decoder.rs

```diff
@@ -265,31 +287,47 @@ impl CallTraceDecoder {
             }
 
             // Decode events
-            self.decode_events(node);
+            self.decode_events(node, &rt);
         }
     }
 
-    fn decode_events(&self, node: &mut CallTraceNode) {
+    fn decode_events(&self, node: &mut CallTraceNode, rt: &RuntimeOrHandle) {
         node.logs.iter_mut().for_each(|log| {
-            self.decode_event(log);
+            rt.block_on(self.decode_event(log));
         });
     }
 
-    fn decode_event(&self, log: &mut RawOrDecodedLog) {
+    async fn decode_event(&self, log: &mut RawOrDecodedLog) {
         if let RawOrDecodedLog::Raw(raw_log) = log {
-            if let Some(events) = self.events.get(&(raw_log.topics[0], raw_log.topics.len() - 1)) {
-                for event in events {
-                    if let Ok(decoded) = event.parse_log(raw_log.clone()) {
-                        *log = RawOrDecodedLog::Decoded(
-                            event.name.clone(),
-                            decoded
-                                .params
-                                .into_iter()
-                                .map(|param| (param.name, self.apply_label(&param.value)))
-                                .collect(),
-                        );
-                        break
-                    }
+            let mut is_external = false;
+
+            let mut events = vec![];
+            if let Some(evs) = self.events.get(&(raw_log.topics[0], raw_log.topics.len() - 1)) {
+                events = evs.clone();
+            } else if let Some(identifier) = &self.signature_identifier {
+                if let Some(event) =
+                    identifier.write().await.identify_event(&raw_log.topics[0].0).await
+                {
+                    events.push(get_indexed_event(event, raw_log));
+                    is_external = true;
+                }
+            }
+
+            for event in events {
+                if let Ok(decoded) = event.parse_log(raw_log.clone()) {
+                    // We want the user to know that this log was decoded using an external
+                    // database, since we just get the first one, and it can be misleading.
+                    let name = if is_external { format!("*{}", event.name) } else { event.name };
```

Comment:
understand the intention, but let's do without the asterisk as it'll cause confusion, Sam's DB is pretty well curated, we can revisit if ppl have issues

```suggestion
                    let name = if is_external { format!("{}", event.name) } else { event.name };
```

Reply:
Hmm I do like the idea of knowing a when it's a "best guess" though and that it may be inaccurate. Alternative ideas to consider:
- Only show the asterisk if a `--verbose-trace` flag is passed. This flag could also show full addresses instead of truncated ones
- At the bottom of a trace, add a "inferred sigs" section that maps selectors to one's looked up in the DB
- Show them as a slightly different shade, such as a lighter blue instead of the regular blue

Personally I lean towards the first option, but definitely not a deal breaker and I don't feel strongly about this